x-pack/plugins/apm/server/lib/helpers/get_document_sources.ts

```diff
@@ -98,49 +156,78 @@ export async function getDocumentSources({
     await apmEventClient.msearch('get_document_availability', ...allSearches)
   ).responses;
 
-  const checkedSources = sourcesToCheck.map((source, index) => {
+  const checkedSources = allSourcesToCheck.map((source, index) => {
+    const { documentType, rollupInterval } = source;
     const responseBefore = allResponses[index * 2];
     const responseAfter = allResponses[index * 2 + 1];
-    const { documentType, rollupInterval } = source;
 
-    const hasDataBefore = responseBefore.hits.total.value > 0;
-    const hasDataAfter = responseAfter.hits.total.value > 0;
+    const hasDocBefore = responseBefore.hits.total.value > 0;
+    const hasDocAfter = responseAfter.hits.total.value > 0;
 
     return {
       documentType,
       rollupInterval,
-      hasDataBefore,
-      hasDataAfter,
+      hasDocBefore,
+      hasDocAfter,
+      checkSummaryFieldExists: source.meta.checkSummaryFieldExists,
     };
   });
 
-  const hasAnyDataBefore = checkedSources.some(
-    (source) => source.hasDataBefore
+  const hasAnySourceDocBefore = checkedSources.some(
+    (source) => source.hasDocBefore
   );
 
-  const sources: Array<ApmDataSource & { hasDocs: boolean }> =
-    checkedSources.map((source) => {
-      const { documentType, hasDataAfter, hasDataBefore, rollupInterval } =
-        source;
+  const sources: TimeRangeMetadata['sources'] = [];
 
-      const hasData = hasDataBefore || hasDataAfter;
+  checkedSources.forEach((source) => {
+    const {
+      documentType,
+      hasDocAfter,
+      hasDocBefore,
+      rollupInterval,
+      checkSummaryFieldExists,
+    } = source;
+    const hasDocBeforeOrAfter = hasDocBefore || hasDocAfter;
+    // If there is any data before, we require that data is available before
+    // this time range to mark this source as available. If we don't do that,
+    // users that upgrade to a version that starts generating service tx metrics
+    // will see a mostly empty screen for a while after upgrading.
+    // If we only check before, users with a new deployment will use raw transaction
+    // events.
+    const hasDocs = hasAnySourceDocBefore ? hasDocBefore : hasDocBeforeOrAfter;
```

Comment:
if you move this up into the .map before, this code might become a lot simpler, e.g.:

```ts
const isSummaryDurationFieldAvailable = Boolean(
      documentType === ApmDocumentType.ServiceTransactionMetric ||
      checkedSources.find(
        (checkedSource) =>
          checkedSource.documentType === documentType &&
          checkedSource.rollupInterval === rollupInterval
      )?.hasDocs);
```